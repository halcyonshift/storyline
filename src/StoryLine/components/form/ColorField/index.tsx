import { useState } from 'react'
import Box from '@mui/material/Box'
import Button from '@mui/material/Button'
import InputLabel from '@mui/material/InputLabel'
import Popover from '@mui/material/Popover'
import Stack from '@mui/material/Stack'
import { TwitterPicker } from 'react-color'
import { useTranslation } from 'react-i18next'
import TooltipIconButton from '@sl/components/TooltipIconButton'
import { GLOBAL_ICONS } from '@sl/constants/icons'
import { ColorType } from '@sl/theme/types'
import { getHex } from '@sl/theme/utils'
import { ColorFieldProps } from './types'

const ColorField = ({ form, name = 'color', label }: ColorFieldProps) => {
    const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null)
    const { t } = useTranslation()
    const colors = [
        'black',
        'white',
        'slate',
        'gray',
        'zinc',
        'neutral',
        'stone',
        'red',
        'orange',
        'amber',
        'yellow',
        'lime',
        'green',
        'emerald',
        'teal',
        'cyan',
        'sky',
        'blue',
        'indigo',
        'violet',
        'purple',
        'fuchsia',
        'pink',
        'rose'
    ].map((color: ColorType) => getHex(color, 400))

    return (
        <Box className='flex'>
            <Stack className='relative flex-grow flex' spacing={1}>
                <InputLabel>{t(label)}</InputLabel>
                <Stack spacing={1} direction='row'>
                    <Button
                        aria-describedby={anchorEl ? `colorfield-${name}` : undefined}
                        className='colorField border w-10 h-10 shadow rounded'
                        sx={{ backgroundColor: form.values[name] || getHex('white') }}
                        onClick={(e) => setAnchorEl(e.currentTarget)}></Button>
                    <TooltipIconButton
                        text='component.colorField.clear'
                        icon={GLOBAL_ICONS.delete}
                        onClick={() => {
                            form.setFieldValue(name, '')
                        }}
                    />
                </Stack>
                {open ? (
                    <Popover
                        id={`colorfield-${name}`}
                        elevation={0}
                        open={Boolean(anchorEl)}
                        anchorEl={anchorEl}
                        onClose={() => setAnchorEl(null)}
                        anchorOrigin={{
                            vertical: 'bottom',
                            horizontal: 'left'
                        }}>
                        <Box className='px-1 py-3'>
                            <TwitterPicker
                                color={form.values[name] || getHex('white')}
                                colors={colors}
                                onChangeComplete={(color) => form.setFieldValue(name, color.hex)}
                            />
                        </Box>
                    </Popover>
                ) : null}
            </Stack>
        </Box>
    )
}

export default ColorField
