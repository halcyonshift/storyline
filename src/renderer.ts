import { init } from '@sentry/electron/renderer'
import { init as reactInit } from '@sentry/react'

if (parseInt(process.env.DEBUG) === 0) {
    init({ dsn: process.env.SENTRY }, reactInit)
}

import '@fontsource/roboto/300.css'
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/500.css'
import '@fontsource/roboto/700.css'
import './index.css'
import './StoryLine'
