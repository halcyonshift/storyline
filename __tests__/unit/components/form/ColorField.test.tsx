import userEvent from '@testing-library/user-event'
import { useFormik } from 'formik'
import renderer from 'react-test-renderer'
import ColorField from '@sl/components/form/ColorField'
import { render, screen } from '../../test-utils'

const mockedUsedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: () => mockedUsedNavigate
}))

const MockColorField = ({ name, label }: { name?: string; label: string }) => {
    const form = useFormik({
        initialValues: { [name]: '#FF0000' },
        onSubmit: async () => {}
    })

    return <ColorField form={form} name={name} label={label} />
}

describe('<ColorField />', () => {
    it('shows the given label', () => {
        render(<MockColorField label='Color Field' />)
        expect(screen.getByText('Color Field')).toBeTruthy()
    })

    it('it clears on click', async () => {
        render(<MockColorField label='Color Field' />)
        await userEvent.click(screen.getAllByRole('button')[1])
        expect(screen.queryByDisplayValue('#FF0000')).toBeFalsy()
    })

    it('it shows popover on click', async () => {
        render(<MockColorField label='Color Field' />)
        expect(screen.queryByRole('presentation')).toBeFalsy()
        await userEvent.click(screen.getAllByRole('button')[0])
        expect(screen.getByRole('presentation')).toBeTruthy()
    })

    it('renders correctly', () => {
        const tree = renderer.create(<MockColorField label='Color Field' />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
