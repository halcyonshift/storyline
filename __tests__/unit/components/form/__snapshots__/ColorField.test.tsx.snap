// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorField /> renders correctly 1`] = `
<div
  className="flex MuiBox-root css-0"
>
  <div
    className="relative flex-grow flex css-1okj3ks-MuiStack-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
    >
      Color Field
    </label>
    <div
      className="css-w4z10b-MuiStack-root"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium colorField border w-10 h-10 shadow rounded css-1veecek-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="component.colorField.clear"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          aria-label="Delete"
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="DeleteOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 9v10H8V9h8m-1.5-6h-5l-1 1H5v2h14V4h-3.5l-1-1zM18 7H6v12c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
